imageloading-core/src/main/java/com/google/accompanist/imageloading/AsyncImage.kt

```diff
@@ -0,0 +1,367 @@
+ * ran in preview mode.
+ * @param shouldRefetchOnSizeChange Lambda which will be invoked when the size changes, allowing
+ * optional re-fetching of the image. Return true to re-fetch the image.
+ */
+@Composable
+fun <R : Any> AsyncImage(
+    state: AsyncImageState<R>,
+    contentDescription: String?,
+    modifier: Modifier = Modifier,
+    alignment: Alignment = Alignment.Center,
+    contentScale: ContentScale = ContentScale.Fit,
+    colorFilter: ColorFilter? = null,
+    fadeIn: Boolean = false,
+    fadeInDurationMs: Int = DefaultTransitionDuration,
+    @DrawableRes previewPlaceholder: Int = 0,
+    shouldRefetchOnSizeChange: (currentResult: ImageLoadState, size: IntSize) -> Boolean = { _, _ -> false },
+) {
+    if (LocalInspectionMode.current && previewPlaceholder != 0) {
+        // If we're in inspection mode (preview) and we have a preview placeholder, just draw
+        // that using an Image and return
+        Image(
+            painter = painterResource(previewPlaceholder),
+            contentDescription = null,
+            modifier = modifier,
+        )
+        return
+    }
+
+    val coroutineScope = rememberCoroutineScope()
+
+    val imageLoader = remember(state, coroutineScope) {
+        ImageLoader(
+            state = state,
+            coroutineScope = coroutineScope,
+            shouldRefetchOnSizeChange = shouldRefetchOnSizeChange,
+        )
+    }
+
+    val cf = if (fadeIn && state.loadState is ImageLoadState.Success) {
+        val fadeInTransition = updateFadeInTransition(
+            key = imageLoader.state,
+            durationMs = fadeInDurationMs
+        )
+        remember { ColorMatrix() }
+            .apply {
+                updateAlpha(fadeInTransition.alpha)
+                updateBrightness(fadeInTransition.brightness)
+                updateSaturation(fadeInTransition.saturation)
+            }
+            .let { matrix ->
+                ColorFilter.colorMatrix(matrix)
+            }
+    } else {
+        // If fade in isn't enabled, just use the provided `colorFilter`
+        colorFilter
+    }
+
+    // Update our ImageLoader with any parameter values changes
+    SideEffect {
+        imageLoader.shouldRefetchOnSizeChange = shouldRefetchOnSizeChange
+    }
+
+    val semantics = if (contentDescription != null) {
+        Modifier.semantics {
+            this.contentDescription = contentDescription
+            this.role = Role.Image
+        }
+    } else Modifier
+
+    // We build a modifier once, for each ImageLoader, which handles everything. We
+    // ensure that no state objects are used in its construction, so that all state
+    // observations are limited to the layout and drawing phases.
+    // We do read some values from the function parameters for the paint() modifier, but they are
+    // limited to less-commonly modified values (alignment + contentScale).
+    Box(
+        modifier
+            .then(semantics)
+            .clipToBounds()
```

Comment:
curious if you really need to clip as we don't clip in our Image implementation. we just draw BitmapPainter with srcBounds which resulted in clipped image

Reply:
Hmmm, `Image` seems to clip: https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/Image.kt;l=183

Reply:
indeed. sounds good then